### ISSUE TITLE
false positive: UMAC_UNCALLABLE_METHOD_OF_ANONYMOUS_CLASS 

### ISSUE THREAD

**Issue by clo-vis** [spotbugs/spotbugs#682]
For each method named _bar_ below, spotbugs issue a warning that the method is "uncallable":

```
class Foo {
    static Runnable x() {
        var k = new Object() {
            void bar() { }
        };
        return k::bar;
    }
    
    static Runnable y() {
        class Bar {
            void bar() { }
        }
        return new Bar()::bar;
    }
    
    static Runnable z() {
        return new Object() {
            void bar() { }
        }::bar;
    }
}
```

**Comment by ThrawnCA** [comment]
OK, technically it's not *uncallable*, but if you're defining an anonymous `Runnable`, why not name it consistently?

    return new Runnable() {
        void run() {}
    }::run;

Note that even without Lambdas, technically any non-private method could be called by retrieving it using reflection. But it's true that these methods cannot be seen and called via the normal procedures.

**Comment by clo-vis** [comment]
It depends on the definition of "normal procedures". Is
```
var x = new Object() {
    private int foo() { return 1; }
};
System.out.println(x.foo());
```
an abnormal procedure?

Additionally, for
```
class Foo {
    static void bar(Iterable<String> it) {
        class LocalClass {  
            int longStrings, stringsContainingUnderscore;
            void countSpecialStrings(String s) { 
                if (s.length() > 1000) longStrings++;
                if (s.indexOf('_') >= 0) stringsContainingUnderscore++;
            }
        }
        LocalClass obj = new LocalClass();
        it.forEach(obj::countSpecialStrings);
        System.out.println(obj.longStrings + "," + obj.stringsContainingUnderscore);
    }
}
```
the warning says _Uncallable method Foo$1LocalClass.countSpecialStrings(String) defined in anonymous class_. 
_LocalClass_ is not an anonymous class, it is a local class.

> A local class is a nested class that is not a member of any class and that has a name.
(Java Language Specification)

The warning goes away if:
- Instead of using a method reference, the method is explicitly invoked, that is, _it.forEach(s -> obj.countSpecialStrings(s))_ instead of _it.forEach(obj::countSpecialStrings)_)
- The method name has only one character

**Comment by ThrawnCA** [comment]
I would say that yes, defining a private method in an anonymous class and then calling it is abnormal. The class is anonymous, what's the point of `private`?

The second issue, where a method reference is not counted the same way as a direct call, is probably worth fixing.

**Comment by rosenthalj** [comment]
I don't know how Spotbugs does its magic,  but I'm not sure this issue is related to issue #811

For issue #811 (java 11 and not java 1.8), Spotbugs is generating  a false positive **UPM_UNCALLED_PRIVATE_METHOD** warning  because it is not identifying that a private class method is actually being called from an anonymous inner class.

When I submitted issue #811 I simplified the problem.  I have a lots of classes with each class having multiple anonymous  inner classes .  Within each of those classes, the inner classes call the same private class method.  Spotbugs is generating a false positive **UPM_UNCALLED_PRIVATE_METHOD** warning for every class that implements this pattern.

My project uses [Apache Wicket 7.x](https://wicket.apache.org/start/wicket-7.x.html) where it is very common (but not unique to Wicket) to have multiple anonymous inner classes within a single class.

**Comment by haumacher** [comment]
This is really no absurd situation, where an anonymous class has a private method used as method reference:
```
	private ProgressDialog createImportProgress() {
                // ...
		return new ProgressDialog(I18NConstants.DOWNLOAD_SQL, px(400), px(300)) {
                        // ...

			@Override
			protected void handleCompleted(DisplayContext context) {
				MessageBox
					.newBuilder(...)
					.buttons(MessageBox.button(..., this::doImport))
					.confirm(...);
			}

			private HandlerResult doImport(DisplayContext context) {
				// ...
			}
		};
	}
```

Here `doImport()` is the implementation of the MessageBox-Button Command interface invoked from a message box button upon completion of the anonymous progress dialog implementation.

The error message here is:
`CORRECTNESS(UMAC_UNCALLABLE_METHOD_OF_ANONYMOUS_CLASS): Uncallable method ....ImportDataDialog$3.doImport(DisplayContext) defined in anonymous class
 at ....ImportDataDialog$3.doImport(ImportDataDialog.java:320)`

**Comment by kstarsinic** [comment]
Here's another related real-world example, names changed to protect the innocent:

```
private static void doSomethingLater() {
  new SwingWorker<Void, Void>() {
    @Override
    protected Void doInBackground() {
      System.out.println("Here I am!");
      return null;
    }

    @Override
    protected void done() {
      System.out.println("All done!");
    }
  }.execute();
}
```
produces "Uncallable method MyClass$1.doInBackground() defined in anonymous class"

**Comment by archiecobbs** [comment]
I'm running into this same false positive bug when using a method reference to a local class.

This is a real bug.

My code looks something like this:

```java
public void someMethod() {

    // this is a local class
    class ValueInfo {

        final AtomicReference<Object> currentValue;
        final Object specialValue;

        void AtomicReference<Object> currentValue() {
            return this.currentValue();
        }

        boolean isValueSpecial() {       // bogus UMAC_UNCALLABLE_METHOD_OF_ANONYMOUS_CLASS reported
            return Objects.equals(this.currentValue.get(), this.specialValue);
        }
    }

    ...

    List<ValueInfo> valueInfos = ...

    ...

    boolean allValuesAreSpecial = Stream.of(valueInfos).allMatch(ValueInfo::isValueSpecial);

    ...

```

**Comment by kilink** [comment]
SpotBugs fails with this warning when writing Gatherers in the style used in the JDK (see [windowFixed](https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/util/stream/Gatherers.java#L87)). I think that is evidence that this is not "abnormal", as code in this style is in the JDK itself, and was added quite recently. Would be great to if this bug were fixed.
